File: ios/RCTOriginalView.m
/*
    Copyright 2024 Audienzz AG

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.

*/

#import "RCTOriginalView.h"
#import "AUConverter.h"
#import "Utils.h"
#import <AudienzziOSSDK/AudienzziOSSDK-Swift.h>

@implementation RCTOriginalView

- (instancetype)init {
    self = [super init];
    
    _backgroundQueue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
    _semaphore = dispatch_semaphore_create(0);
    
    return self;
}

- (void)didSetProps:(NSArray<NSString *> *)value {
    if (_propsChanged) {
        dispatch_async(self.backgroundQueue, ^{
            [self createAd];
        });
    }
    _propsChanged = NO;
}

- (void)setPlaybackMethod:(NSArray<NSString *> *)value {
    _playbackMethod = value;
    _propsChanged = YES;
}

- (void)setIsLazyLoad:(NSNumber *)value {
    _isLazyLoad = [value boolValue];
    _propsChanged = YES;
}

- (void)setAdFormats:(NSArray<NSString *> *)value {
    _adFormats = value;
    _propsChanged = YES;
}

- (void)setApiParameters:(NSArray<NSString *> *)value {
    _apiParameters = value;
    _propsChanged = YES;
}

- (void)setVideoProtocols:(NSArray<NSString *> *)value {
    _videoProtocols = value;
    _propsChanged = YES;
}

- (void)setVideoBitrate:(NSArray<NSNumber *> *)value {
    _videoBitrate = value;
    _propsChanged = YES;
}

- (void)setVideoDuration:(NSArray<NSNumber *> *)value {
    _videoDuration = value;
    _propsChanged = YES;
}

- (void)setKeyword:(NSString *)value {
    _keyword = value;
    _propsChanged = YES;
}

- (void)setKeywords:(NSArray<NSString *> *)value {
    _keywords = value;
    _propsChanged = YES;
}

- (void)setAppContent:(NSDictionary *)value {
    AUMORTBAppContent *appContent = [Utils createContentObjectFromDictionary:value];
    _appContent = appContent;
    _propsChanged = YES;
}

- (void)setPbAdSlot:(NSString *)value {
    _pbAdSlot = value;
    _propsChanged = YES;
}

- (void)setGpID:(NSString *)value {
    _gpID = value;
    _propsChanged = YES;
}

- (void)setAuConfigID:(NSString *)value {
    _auConfigID = value;
    _propsChanged = YES;
}

- (void)setAdUnitID:(NSString *)value {
    _adUnitID = value;
    _propsChanged = YES;
}

- (void)createAd {}

- (void)internalCreateAd {
    NSMutableArray<AUVideoPlaybackMethod *> *playbackMethodArray = [NSMutableArray array];
    NSMutableArray<AUApi *> *apiArray = [NSMutableArray array];
    NSMutableArray<AUVideoProtocols *> *protocolArray = [NSMutableArray array];
    
    _bannerParameters = [[AUBannerParameters alloc] init];
    _videoParameters = [[AUVideoParameters alloc] initWithMimes:@[@"video/mp4"]];
    
    for (NSString *playbackMethodString in _playbackMethod) {
        AUVideoPlaybackMethod *playbackMethod = [AUConverter convertToAUPlaybackMethod:playbackMethodString];
        if (playbackMethod) {
            [playbackMethodArray addObject:playbackMethod];
        }
    }
    for (NSString *apiString in _apiParameters) {
        AUApi *api = [AUConverter convertToAUApi:apiString];
        if (api) {
            [apiArray addObject:api];
        }
    }
    for (NSString *protocolString in _videoProtocols) {
        AUVideoProtocols *protocol = [AUConverter convertToAUProtocols:protocolString];
        if (protocol) {
            [protocolArray addObject:protocol];
        }
    }
    
    _bannerParameters.api = [apiArray copy];
    _videoParameters.api = [apiArray copy];
    _videoParameters.protocols = [protocolArray copy];
    _videoParameters.playbackMethod = [playbackMethodArray copy];
    [_videoParameters setMinBitrate:_videoBitrate[0]];
    [_videoParameters setMaxBitrate:_videoBitrate[1]];
    [_videoParameters setMinDuration:_videoDuration[0]];
    [_videoParameters setMaxDuration:_videoDuration[1]];
}

@end

File: android/src/main/java/com/audienzzrn/RCTOriginalView.kt
package com.audienzz

/*
    Copyright 2024 Audienzz AG

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.

*/

import android.content.Context
import android.widget.FrameLayout
import org.audienzz.mobile.AudienzzBannerParameters
import org.audienzz.mobile.AudienzzSignals
import org.audienzz.mobile.AudienzzVideoParameters

open class RCTOriginalView(context: Context) : FrameLayout(context) {
  var adFormats: List<String> = listOf()
  var adUnitID: String = ""
  var auConfigID: String = ""
  var videoBitrate: List<Int>? = null
  var videoDuration: List<Int>? = null
  var pbAdSlot: String? = null
  var gpID: String? = null
  var keyword: String? = null
  var playbackMethods: List<String> = listOf()
  var apiParameters: List<String> = listOf()
  var videoProtocols: List<String> = listOf()
  var isLazyLoad: Boolean = true
  var isAdaptive: Boolean = false
  var impOrtbConfig: String? = null
  private var propsChanged: Boolean = false

  var bannerParameters = AudienzzBannerParameters()
  var videoParameters = AudienzzVideoParameters(listOf("video/x-flv", "video/mp4"))

  open fun createAd() {
    val allowedApis = mutableListOf<AudienzzSignals.Api>()
    val allowedPlaybackMethods = mutableListOf<AudienzzSignals.PlaybackMethod>()
    val allowedProtocols = mutableListOf<AudienzzSignals.Protocols>()

      for (apiString in apiParameters) {
        val api = AudienzzConversionUtils.convertToAudienzzApi(apiString)
        if (api != null) {
          allowedApis.add(api)
        }
      }
      for (protocolString in videoProtocols) {
        val protocol = AudienzzConversionUtils.convertToAudienzzVideoProtocols(protocolString)
        if (protocol != null) {
          allowedProtocols.add(protocol)
        }
      }
      for (playbackMethodString in playbackMethods) {
        val playbackMethod =
          AudienzzConversionUtils.convertToAudienzzPlaybackMethod(playbackMethodString)
        if (playbackMethod != null) {
          allowedPlaybackMethods.add(playbackMethod)
        }
      }

    bannerParameters.api = allowedApis

    videoParameters.api = allowedApis
    videoParameters.protocols = allowedProtocols
    videoParameters.playbackMethod = allowedPlaybackMethods
    videoParameters.placement = AudienzzSignals.Placement.Interstitial
    if (videoDuration != null) {
      videoParameters.minDuration = videoDuration!![0]
      videoParameters.maxDuration = videoDuration!![1]
    }
    if (videoBitrate != null) {
      videoParameters.minBitrate = videoBitrate!![0]
      videoParameters.maxBitrate = videoBitrate!![1]
    }

  }

  fun updatePbAdSlot(value: String) {
    pbAdSlot = value
  }

  fun updateGpID(value: String) {
    gpID = value
  }

  fun updateAdFormats(value: List<String>) {
    adFormats = value
  }

  fun updateApiParameters(value: List<String>) {
    apiParameters = value
  }

  fun updateVideoProtocols(value: List<String>) {
    videoProtocols = value
  }

  fun updatePlaybackMethod(value: List<String>) {
    playbackMethods = value
  }

  fun updateVideoBitrate(value: List<Int>) {
    videoBitrate = value
  }

  fun updateVideoDuration(value: List<Int>) {
    videoDuration = value
  }

  fun updateAuConfigID(value: String) {
    auConfigID = value
  }

  fun updateAdUnitID(value: String) {
    adUnitID = value
  }

  fun updateImpOrtbConfig(value: String){
    impOrtbConfig = value
  }

  fun updateIsLazyLoad(value: Boolean) {
    isLazyLoad = value
  }

  fun updateIsAdaptive(value: Boolean) {
    isAdaptive = value
  }

  fun updatePropsChanged(value: Boolean) {
    propsChanged = value
  }

  fun getPropsChanged(): Boolean {
    return propsChanged
  }
}
